pennylane_lightning_gpu/src/simulator/MPIWorker.hpp

```diff
@@ -0,0 +1,284 @@
+                       const size_t numLocalQubits, cudaStream_t localStream) {
+
+    custatevecSVSwapWorkerDescriptor_t svSegSwapWorker = nullptr;
+
+    int nDevices_int = 0;
+    PL_CUDA_IS_SUCCESS(cudaGetDeviceCount(&nDevices_int));
+
+    size_t nDevices = static_cast<size_t>(nDevices_int);
+
+    // Ensure the number of P2P devices is calculated based on the number of MPI
+    // processes within the node
+    nDevices = mpi_manager.getSizeNode() < nDevices ? mpi_manager.getSizeNode()
+                                                    : nDevices;
+
+    size_t nP2PDeviceBits = std::bit_width(nDevices) - 1;
+
+    cudaDataType_t svDataType;
+    if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                  std::is_same_v<CFP_t, double2>) {
+        svDataType = CUDA_C_64F;
+    } else {
+        svDataType = CUDA_C_32F;
+    }
+
+    cudaEvent_t localEvent = nullptr;
+    custatevecCommunicatorDescriptor_t communicator = nullptr;
+
+    PL_CUDA_IS_SUCCESS(cudaEventCreateWithFlags(
+        &localEvent, cudaEventInterprocess | cudaEventDisableTiming));
+
+    custatevecCommunicatorType_t communicatorType;
+    if (mpi_manager.getVendor() == "MPICH") {
+        communicatorType = CUSTATEVEC_COMMUNICATOR_TYPE_MPICH;
+    }
+    if (mpi_manager.getVendor() == "Open MPI") {
+        communicatorType = CUSTATEVEC_COMMUNICATOR_TYPE_OPENMPI;
+    }
+
+    auto err = custatevecCommunicatorCreate(handle, &communicator,
+                                            communicatorType, nullptr);
+    if (err != CUSTATEVEC_STATUS_SUCCESS) {
+        communicator = nullptr;
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecCommunicatorCreate(
+            handle, &communicator, communicatorType, "libmpi.so"));
+    }
+    mpi_manager.Barrier();
+
+    void *d_extraWorkspace = nullptr;
+    void *d_transferWorkspace = nullptr;
+
+    std::vector<void *> d_subSVsP2P;
+    std::vector<int> subSVIndicesP2P;
+    std::vector<cudaEvent_t> remoteEvents;
+
+    size_t extraWorkspaceSize = 0;
+    size_t minTransferWorkspaceSize = 0;
+
+    PL_CUSTATEVEC_IS_SUCCESS(custatevecSVSwapWorkerCreate(
+        /* custatevecHandle_t */ handle,
+        /* custatevecSVSwapWorkerDescriptor_t* */ &svSegSwapWorker,
+        /* custatevecCommunicatorDescriptor_t */ communicator,
+        /* void* */ sv,
+        /* int32_t */ mpi_manager.getRank(),
+        /* cudaEvent_t */ localEvent,
+        /* cudaDataType_t */ svDataType,
+        /* cudaStream_t */ localStream,
+        /* size_t* */ &extraWorkspaceSize,
+        /* size_t* */ &minTransferWorkspaceSize));
+
+    PL_CUDA_IS_SUCCESS(cudaMalloc(&d_extraWorkspace, extraWorkspaceSize));
+
+    PL_CUSTATEVEC_IS_SUCCESS(custatevecSVSwapWorkerSetExtraWorkspace(
+        /* custatevecHandle_t */ handle,
+        /* custatevecSVSwapWorkerDescriptor_t */ svSegSwapWorker,
+        /* void* */ d_extraWorkspace,
+        /* size_t */ extraWorkspaceSize));
+
+    size_t transferWorkspaceSize;
+    if (mpi_buffer_size == 0) {
```

Comment:
Why not sync the two definitions as log2 of the number of bytes? `mpi_buffer_size` could then be an optional argument with a default value of `26`.

Reply:
If `mpi_buffer_size == 0` we are using a default value 26, we should define this number elsewhere I think, which yields a `transferWorkspaceSize` of `2**26`. Otherwise, we use `mpi_buffer_size` to define a `transferWorkspaceSize` of `2**(mpi_buffer_size)* sizeof(double) * 2`. I think it we should have the same buffer size if `mpi_buffer_size == 26`. `mpi_buffer_size` can refer to log2 of the number of bytes, or the number of qubits. If this is exposed to the user, it may be more intuitive to use number of qubits.